### DISCUSSION TITLE
ECS Configuration

### DISCUSSION THREAD

**Discussion by LiamJBerridge** [artilleryio/artillery#2280] [Artillery]
Hey, 

We're running tests using fargate and we're seeing memory and cpu maxing out at around 100% which is then causing the tests to timeout rather than fail. 

Below are the 2 commands that we are trying to run with values based on https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-cpu-memory-error.html

artillery run-fargate --config config.yml scenarios/initialiseHomepage.yml --dotenv "../.env" --region eu-west-1 --launch-config '{"cpu": 8192, "memory": 32768, "ulimits":[{"name":"nofile","softLimit":"32768","hardLimit":"32768"}]}'
Error: Unsupported cpu override for Fargate. Must be one of: 256, 512, 1024, 2048, 4096

artillery run-fargate --config config.yml scenarios/initialiseHomepage.yml --dotenv "../.env" --region eu-west-1 --launch-config '{"cpu": 4096, "memory": 29000, "ulimits":[{"name":"nofile","softLimit":"32768","hardLimit":"768"}]}
Error: Fargate memory override for cpu = 4096 must be one of: 8192, 9216, 10240, 11264, 12288

We can get it to work with memory upto 16gb but past that it errors on the CPU memory. Has anyone managed to push these values higher successfully?

**Comment by hassy** [comment]
Hi @LiamJBerridge 👋 That's a bug in Artillery itself, we need to update the allowed CPU/memory configurations (as listed here: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-cpu-memory-error.html). At the time we added Fargate support 12GB was the maximum allowed amount of RAM for a 4 vCPU container.

**Reply by hassy** [reply]
Will fix in the upcoming release (this Friday)

**Reply by hassy** [reply]
In the meantime you should be able to reduce the number of VUs per container, and scale out horizontally instead (with `--count`)

**Reply by LiamJBerridge** [reply]
Hey @hassy 👋 

Was there a release on friday, I can't see an update in NPM?

**Reply by hassy** [reply]
There wasn't, my apologies! Aiming for this Friday now.

**Reply by LiamJBerridge** [reply]
👋 

No worries, can you confirm something for me? 

MaxVUsers? 

If you're running on fargate will that limit the the vusers per task or is it per phase? We're seeing an issue with horizontal scaling where if you invoke 32 tasks for example, 4 will run at 100% and the other 28 will run idle. 

What we see is that the 4 that are running are timing out trying to launch browsers but there are 28 tasks that could take on the load?

However, I've seen a few different definitions for the param in the documentation?

Cheers 🍻

**Reply by LiamJBerridge** [reply]
@hassy 

Here is an example 

![Screenshot 2023-11-15 at 11 50 32](https://github.com/artilleryio/artillery/assets/45438533/900fe481-e1e9-431a-98b6-cf8fc15705bb)
![Screenshot 2023-11-15 at 11 51 24](https://github.com/artilleryio/artillery/assets/45438533/de6c8cc1-813c-4d62-9531-84479828237a)

You can see that there are a lot of workers sat idle when running the test with a --count. I'd expect each task to take an equal proportion of the load but it looks like a small number take all of the load?

**Reply by hassy** [reply]
`maxVusers` is per task, so `--count` acts as a multiplier of what's defined in `config.phases` in the script, e.g. if you have a phase with `arrivalRate: 10`, and run with `--count 10`, you'll see 100 VUs created every second in total. Those idling tasks do look odd, you'd expect to see all of them running the test script and show similar loads.

**Reply by bernardobridge** [reply]
@LiamJBerridge, when you run this again, can you check Cloudwatch logs for any of the idling tasks and see if there's any error or mention from inside them? 

I think the best would be to look at the logs and try to see if there was either an error starting them (i.e. AWS error), or within the workers themselves.

**Reply by hassy** [reply]
Fix for this incoming in https://github.com/artilleryio/artillery/pull/2301
